New error api

cgwalters opened the pull request.

cgwalters commented:
See also https://github.com/flatpak/flatpak/blob/b00b8b1601142d21564a2dff871b8750db2af71d/common/flatpak-utils.c#L288

cgwalters commented:
@alexlarsson Do you like ⬆️ ?

alexlarsson commented:
I like it. I have the same in flatpak, called flatpak_fail(error, "blah...")

alexlarsson commented:
See https://github.com/flatpak/flatpak/blob/master/common/flatpak-utils.c

jlebon commented on glnx-errors.h:
```diff
@@ -25,25 +25,65 @@
+  int errsv = errno;
+  g_set_error_literal (error, G_IO_ERROR,
+                       g_io_error_from_errno (errsv),
+                       g_strerror (errsv));
```
You still have to restore `errno` here, right? (And similarly in `glnx_throw_errno_prefix()`).

cgwalters replied:
I'm not sure anything was actually relying on that...but it can't hurt I guess?  

Going back to [the original commit](https://git.gnome.org/browse/libgsystem/commit/?id=ed106741f7a0596dc8b960b31fdae671d31d666d) I don't see why I did that.

cgwalters commented:
Anyways it can't hurt, fixup ⬆️ .  I added comments so the person wondering why it's there later may decide to delete it.

gnomesysadmins commented:
Thank you for contributing to libglnx!

libglnx uses Bugzilla for code review.

If you have never contributed to GNOME before make sure you have read the
getting started documentation:
http://www.gnome.org/get-involved

Otherwise please visit
https://wiki.gnome.org/Newcomers/CodeContributionWorkflow
and follow the instructions there to upload your change to Bugzilla.

jlebon commented on glnx-errors.h:
```diff
@@ -25,25 +25,101 @@
+  va_start (args, fmt);
+  GError *new = g_error_new_valist (G_IO_ERROR, G_IO_ERROR_FAILED, fmt, args);
+  va_end (args);
+  g_assert_no_error (*error);
```
Sorry, don't mean to nitpick too much. :) Should we really make this an assertion though? Feels like we should just use `g_propagate_error` here, no? It warns about overwriting the error, but leaves it untouched.

cgwalters replied:
For something like this I think it deserves careful review!  That's a good point.  I originally did the assertion because I only was looking at `g_set_error()` as handling the overwritten warning, but you're definitely right, propagate should work.  Will do.

cgwalters commented:
Okay, now that this PR is closed GH won't render updates.  See [my branch](https://github.com/cgwalters/libglnx/tree/new-error-api) where I've added a few fixups, and also added tests ✅.

jlebon commented:
👍 This is really nice!

cgwalters commented:
Thanks!  Pushed 🚅
